controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java

```diff
@@ -944,118 +944,158 @@ private boolean isSubscriberToNonSubscriberTransition(final ReaderGroupConfigRec
-                            .thenApply(gt -> {
-                            if (gt) {
-                                return maxRecord;
-                              } else {
-                                return lowerBound;
-                              }
-                            });
-                            })).orElse(CompletableFuture.completedFuture(null));
+                        } else { // greater than max. truncate at max. 
+                            // let there be data loss. its a lagging reader.. but if there is no streamcut in 
+                            // retentionset that satisfied the min and max criteria, then we should simply truncate at 
+                            // least at the lowerbound 
+                            return Optional.ofNullable(limits.getKey())
+                                           .filter(x -> currentSize - x.getRecordingSize() < retainedSizeLB)
+                                           .map(x -> streamMetadataStore.getStreamCutRecord(scope, stream, x, context, executor)
+                                                                        .thenApply(StreamCutRecord::getStreamCut))
+                                    .orElse(CompletableFuture.completedFuture(lowerBound));
                         }
-                      }
+                    }
                 });
     }
 
     private CompletableFuture<Map<Long, Long>> getTruncationStreamCutByTimeLimit(String scope, String stream, OperationContext context,
                                                                                  RetentionPolicy policy, RetentionSet retentionSet,
                                                                                  Map<Long, Long> lowerBound) {
-        Map.Entry<StreamCutReferenceRecord, StreamCutReferenceRecord> limits =
-                getBoundStreamCuts(policy, retentionSet);
+        long currentTime = retentionClock.get().get();
+
+        Map.Entry<StreamCutReferenceRecord, StreamCutReferenceRecord> limits = getBoundStreamCuts(policy, retentionSet, 
+                x -> currentTime - x.getRecordingTime());
         // if subscriber lowerbound is ahead of streamcut corresponding to the max time and is behind stream cut for min time
         // from the retention set then we can safely truncate at lowerbound. Else we will truncate at the max time bound if it
         // exists
-        // 1. if LB > min => truncate at min
-        // 2. if LB < max => truncate at max
-        // 3. if LB < min && LB > max => truncate at LB
-        // 4. if LB < min && overlaps max => truncate at max
+        // 1. if LB is ahead of min => truncate at min
+        // 2. if LB is behind max => truncate at max
+        // 3. if LB is behind min && LB is ahead of max => truncate at LB
+        // 4. if LB is behind min && overlaps max => truncate at max
+        // 5. if LB overlaps with min and max ==> so its got both recent data and older data.. what do we do.. 
```

Comment:
Here we should truncate at a stream-cut that would make the stream size closest to but **lesser than** `max` post truncation. where are the 2 options?

Reply:
the next line in the comment gives out two options:
        //      there are two options: either truncate beyond max or not truncate at all.  

and we we truncate beyond max. i have updated the comment to reflect that.